wiki/Writing-tests-with-good-behavioural-coverage.md

```diff
@@ -0,0 +1,1265 @@
+    passbookFile.createNewFile()
+    val output = ByteArrayOutputStream()
+    System.setOut(PrintStream(output))
+
+    account.withdraw("user", "password", 200.0, file = passbookFile)
+
+    // Read the passbook file and check its contents
+    val fileContents = passbookFile.readText()
+    assertThat(fileContents).contains("Withdrew 200.0. New balance is 800.0")
+
+    System.setOut(System.out)
+}
+```
+
+**d. To verify based on the log message:**
+
+Validates that the correct message about the withdrawal is logged.
+
+```
+@Test
+fun testWithdraw_withSufficientBalance_logsCorrectMessage() {
+    val output = ByteArrayOutputStream()
+    System.setOut(PrintStream(output))
+
+    account.withdraw("user", "password", 200.0)
+
+    assertThat(output.toString().trim()).contains("New balance is 800.0")
+    System.setOut(System.out)
+}
+```
+
+**e. To verify with the balance is updated correctly:**
+
+Ensures that the balance is updated correctly after a withdrawal.
+
+```
+@Test
+fun testWithdraw_withSufficientBalance_updatesBalanceCorrectly() {
+    account.withdraw("user", "password", 200.0)
+    assertThat(account.balance).isEqualTo(800.0)
+}
+```
+
+These tests cover various aspects of the withdrawal functionality, ensuring that the balance updates correctly with receipts, passbooks and output messages. Although the core functionality of withdrawing funds and updating the balance is consistent, it can be observed in multiple ways. Each test focuses on a specific verification method while ultimately validating the same core functionality.
+
+## Testing Practices for Effective Validation
+
+The following best practices are meant to assist in the organization and execution of tests. By following these guidelines, you can ensure that your testing process is efficient, thorough, and well-structured.
+
+### 1. Use of Helper Functions
+
+Helper functions are crucial for avoiding repetitive code, especially for pre-setup tasks that are used across multiple test cases. In this scenario, initializing the bank data or ensuring the proper state of the BankAccount object can be encapsulated in helper functions.
+
+```
+// Helper function to initialize a BankAccount with default data
+fun createDefaultBankAccount(): BankAccount {
+    return BankAccount(1000.0, "user", "password")
+}
+```
+
+This helps to reduce redundancy and maintain consistency across multiple test cases by centralizing repetitive setup tasks.
+
+### 2. Setup and Teardown using `@Before` and `@After`
+
+Using `@Before` and `@After` annotations ensures that common setup and cleanup tasks are automatically executed before and after each test case, maintaining a clean and consistent test environment.
+
+**Example:**
+
+```
+class BankAccountTests {
+
+    private lateinit var account: BankAccount
+
+    @Before
+    fun setUp() {
+        // Initialize a BankAccount instance before each test
+        account = createDefaultBankAccount()
+    }
+
+    @After
+    fun cleanup() {
```

Comment:
Nit: cleanUp (verb form), but perhaps tearDown is better given the section name?

Reply:
Yeah, while naming it as `restoreStreams` was followed for restoring system outputs in various places, 

```
@After
fun restoreStreams() {
  System.setOut(originalOut)
}
```

thought `cleanUp` would provide a more generalized name, but `tearDown` is the standard practice across the codebase, will stick with that and updated it to use `tearDown`. Thanks.